README.md

```diff
@@ -1824,4 +1824,134 @@ When you are dealing with exceptions in your code, use `throw new RuntimeExcepti
+
+### Exporting Environment Variables to Sub Shells
+
+Sub shells can be created in the command line using the command `bash`. By default, these sub shells will not have access to (inherit) the variables of the shell they were created from.
+
+In the following example, a new variable called `a` is created and assigned the value of `learning-bash.sh` (line 1). The command to print the value of `a` is line 2 and the value of `a` is line 3. Line 4 creates a new sub shell. Line 5 calls print on value `a` and line 6 returns the value of `a` in the sub shell.
+
+```
+1    a=learning-bash.sh
+2    echo $a
+3    learning-bash.sh
+4    bash
+5    echo $a
+6    
+```
+
+The following example begins in the same way as the one above, but just before the new sub shell is created (`bash`) we can use the command `export` followed by the name of the variable to extend the value of this variable to the sub shell. Then when we print the value of `a` (`echo $a`), the value of `a` defined in the original shell has been inherited by the sub shell.
+
+```
+1    a=learning-bash.sh
+2    echo $a
+3    learning-bash.sh
+4    export a
+5    bash
+6    echo $a
+7    learning-bash.sh
+```
+
+Note: In order for the value of `a` to be available in the sub shell, the sub shell must be created just after the variable has been exported.
+
+## `finally`
+
+The `finally` block in Java can be useful for making sure specified parts of code are run last, but it is worth using carefully as it can also assist with unintentionally overriding previous commands.
+
+The following example contains a try-with-resources statement. This is when a `try` statement declares one or more resources (`BufferedWriter`). Resources are objects that must be closed after the program is finished with them i.e. at the end of the statement.
+
+Because `BufferedWriter` is a resource that needs to be closed, once the `writer.write()` method has been called an `IOException` is automatically thrown.
+
+``` java
+ @Test
+    public void whenWriteStringUsingBufferedWriter_thenCreateFile() {
+        String home = System.getProperty("user.home");
+        String json = "{}";
+        try {
+            writer = new BufferedWriter(new FileWriter(home + "/.config/budjen/transactions.json"));
+            writer.write(json);
+            } catch (IOException e) {
+            throw new RuntimeException(e);
+        } finally {
+            try {
+                if (writer != null) {
+                    writer.close();
+                }
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+        }
+    }
+```
+
+It is tempting to use a `finally` block to make sure the resource `BufferedWriter` will be closed even when an exception is thrown, but using a `finally` block in this case allows the new exception to override the previous exception so the contents of that will be lost.
```

Comment:
In fact, prior to Java 7 (I think, it might have been 6) you *had* to `close()` resources in a finally to avoid resource leaks. So you might still encounter this valid approach if you look at code that's old enough.

Reply:
It looks like `java.io.Closeable` was introduced in Java 5:
https://docs.oracle.com/javase/7/docs/api/java/io/Closeable.html
Does this mean a finally block in the catch block is valid before Java 5?